### ISSUE TITLE
Improve statuses and initialization state

### ISSUE THREAD

**Issue by nickofthyme** [elastic/elastic-charts#937]
I would like to start a chain to discuss the way we handle the chart status in the `ChartContainer` component.

Currently, we assign one of the following statues to the chart based on some pretty simple logic.

```ts
export const InitStatus = Object.freeze({
  ParentSizeInvalid: 'ParentSizeInvalid' as const,
  SpecNotInitialized: 'SpecNotInitialized' as const,
  MissingChartType: 'MissingChartType' as const,
  ChartNotInitialized: 'ChartNotInitialized' as const,
  Initialized: 'Initialized' as const,
});
```

However, these are not necessarily always entirely accurate. For example, the chart is initialized but has no series defined, results in `MissingChartType`. Furthermore, this `InitStatus` seems to merge the idea of the chart status with chart errors in processing. I am proposing we refactor some of these statuses to flag errors, warnings and status separately.

Below are the proposed statuses and errors

### Statuses
- `NotInitialized` - Starting status before any spec is upserted
- `Processing` - Processing (aka upserting) specs
- `Initialized` - Chart has processed all specs
- `Rendering` - Starting to draw on canvas
- `Rendered` - Chart rendered (not sure if this should be here or elsewhere)

> I think updates to this status could be debounced so that simple charts don't update status that are milliseconds apart.

### Errors
> These should be fatal errors that we are unable to correct or find a suitable fallback case
- `MissingChartType` - _(Not sure what this is for)_
- `ParentSizeInvalid` - Bad dimensions

### Warnings
> These have a suitable fallback case, possibly multiple at once
- All current `Logger.warn` calls

Thoughts? 🤔 cc: @markov00 @monfera @rshen91

**Comment by monfera** [comment]
What are our goals with these status indicators?

Some questions:
- what's the name of the status of the chart when it "exists" but it's not initialized yet? I guess it's when it's `NotInitialized` but it'd be useful to have a name based on what it already is (eg. `Instantiated` or similar) - hmm maybe it's a rename suggestion
- under what circumstances would anyone (our code, or userland code) read a `Processing` state, assuming that spec upsert is synchronous? It'd be invisible to the outside, so what purpose does it serve in our internal processing?
- same with Rendering vs Rendered

Also, while the description mentions the current statuses aren't always accurate, what's the driver for even having statuses? I feel like every status needs some justification, as state management is tedious.

Especially if we introduce asynchrony, debounce and such.

Also I'm not sure what the status would be in combined cases, eg. when creating a chart with a simultaneous specification of the spec.

So a little table would be useful as it is, with the status name paired with the 
- possible users (readers) of that state: chart internal or external (API), or both?
- why is this state important for whoever reads the status? what's the purpose?
- what alternatives exist to the creation of this state, to still meet the previous purpose?

Maybe we could chat about state in general; I'm already a bit uncomfortable with the current way state works, because for us to pretend the chart is a React component, ie. allow arbitrary specification in JSX, already makes us go through some hoops, and I'm not sure if we have solid reasons for using JSX instead of JavaScript objects, if I understand the latter could be similarly structured but wouldn't involve state management tedium. I may overlook something there.

Also, I'm not sure how the status indicators help consistent error handling (#117) 

Is a chart loading indicator status (#202) an actual need currently? The charts get the data from the user app and we have no idea if the user app is waiting for the data. In theory we can add an overlay (progress indicator) that a client can asynchronously update every once in a while, and we'd display this, but it doesn't seem to relate to our internal chart state. I don't think it's about a need to draw a progress bar _while_ we data process or render data, as we're synchronous.

Regarding status indicators, it's also annoying for a dashboard or canvas page to have 5-10 charts, all or many of which sport a flashing progress indicator while loading something. For this reason, we removed the spinner from Canvas charts, and put a progress indicator to the top level of the Canvas page. So it feels more like client responsibility.

Not sure how it relates to documentation (#379) either.

Let's discuss it on a call, also in part b/c there's a ton of other types of work to do, so it's good to see which item helps how.

I may oversimplify here, but as chart data processing and rendering are synchronous and they don't take a lot of time. I think of a chart - whether it uses SVG, HTML, Canvas2d or WebGL as the substrate for markers etc. - as a simple React component that 
- internally goes through plain lifecycle events eg. `componentDidMount`, but it's an implementation detail
- if there's justified need, it might call some user app callback associated with such lifecycle events
- may provide user app callback functions (or other means, eg. rerendering with changed props) that enable the user app to add/remove series, data points, invoke some interactions eg. show the hover bar or tooltip etc.
- may call some user app callback associated with user interactions such as what the end user is hovering over, clicking on, box selects etc.

These don't seem to require status flags.

I'd also like to talk with you all about incremental data processing for the future, eg. the data arrives from ES in chunks, and the user app passes us data in chunks too.

**Comment by nickofthyme** [comment]
Thanks for the thorough response. I think we table this discussion for now and revisit it in the future.

**Comment by markov00** [comment]
With the near future advent of the new reconciler, we don't need anymore these type of states, or at least we can remove most of them

**Comment by nickofthyme** [comment]
What if instead of using a custom reconciler we switch to a json config style api instead of a react component api? Just thinking out-loud. This has the following benefits in my eyes:

- Framework agnostic 
- Easier typings for conditional types across all config options
- ...

**Comment by monfera** [comment]
@nickofthyme 

> What if instead of using a custom reconciler we switch to a json config style api instead of a react component api?

Agree, and also have asked for this change, including for one of your reasons—one fewer coupling to an opinionated dependency. IIRC the original motivation with the React-ish API was to show a sense of continuity with `uber/react-vis` which was initially expected of Marco.

The upsert loop yields a non-idiomatic data and control flow, which also prevents some type simplifications, and ossifies the front-end. Eg. its added friction prevented me from completing a timeboxed PoC for heterogeneous charts in one container. One of the several inconveniences is that even though the API user passes on the entire chart spec, it's processed incrementally, although some of our logic requires the info from all spec elements. The non-essential complexity makes the charts library less approachable to onboarding developers.

Had some chats with @markov00 earlier, including going with JSON in the API, my understanding was that switching to JSON is mostly a question of time and development capacity

**Comment by nickofthyme** [comment]
> the JSON config doesn't preclude the ability to have a react bridge

That's true, but to Robert's point it does inherently create constraints, in addition to rework.

> my understanding was that switching to JSON is mostly a question of time and development capacity

Ok great! Should we open a new issue solely dedicated to the discussion of migrating to JSON schema? This issue is not really germane to the json or react reconciler idea.

**Comment by markov00** [comment]
>Ok great! Should we open a new issue solely dedicated to the discussion of migrating to JSON schema? This issue is not really germane to the json or react reconciler idea.

Sure, I think we can start discussing that
